server/pulp/server/async/app.py

```diff
@@ -30,6 +31,92 @@
 _logger = logging.getLogger(__name__)
 
 
+class HeartbeatStep(bootsteps.StartStopStep):
+
+    """
+    This class is a celery "Blueprint". It extends the functionality of the celery
+    worker by establishing a timer on worker startup which calls the '_record_heartbeat()'
+    method periodically. This allows each worker to write its own worker record to the
+    database, instead of relying on pulp_celerybeat to do so.
+
+    http://docs.celeryproject.org/en/master/userguide/extending.html
+    https://groups.google.com/d/msg/celery-users/3fs0ocREYqw/C7U1lCAp56sJ
+
+    :param worker: The worker instance (unused)
+    :type  worker: celery.apps.worker.Worker
+    """
+
+    requires = ('celery.worker.components:Timer', )
+
+    def __init__(self, worker, **kwargs):
+        """
+        The step init is called when the worker instance is created, It is called with the
+        worker instance as the first argument and all keyword arguments from the original
+        worker.__init__ call.
+
+        :param worker: The worker instance (unused)
+        :type  worker: celery.apps.worker.Worker
+        """
+        self.timer_ref = None
+
+    def start(self, worker):
+        """
+        This method is called when the worker starts up and also whenever the AMQP connection is
+        reset (which triggers an internal restart). The timer is reset when the connection is lost,
+        so we have to install the timer again for every call to self.start.
+
+        :param worker: The worker instance
+        :type  worker: celery.apps.worker.Worker
+        """
+        self.timer_ref = worker.timer.call_repeatedly(
+            constants.PULP_PROCESS_HEARTBEAT_INTERVAL,
+            self._record_heartbeat,
+            (worker, ),
+            priority=10,
+        )
+        self._record_heartbeat(worker)
+
+    def stop(self, worker):
+        """
+        This method is called every time the worker is restarted (i.e. connection is lost)
+        and also at shutdown. The worker calls stop() at shutdown only.
```

Comment:
The last sentence here is confusing. Can this be checked for clarity/correctness.